dsls/sql/src/main/java/org/apache/beam/dsls/sql/BeamSql.java

```diff
@@ -0,0 +1,292 @@
+ * {@link PTransform}, so developers can use standard SQL queries in a Beam pipeline.
+ *
+ * <h1>Beam SQL DSL usage:</h1>
+ * A typical pipeline with Beam SQL DSL is:
+ * <pre>
+ *{@code
+PipelineOptions options = PipelineOptionsFactory.create();
+Pipeline p = Pipeline.create(options);
+
+//create table from TextIO;
+TableSchema tableASchema = ...;
+PCollection<BeamSqlRow> inputTableA = p.apply(TextIO.read().from("/my/input/patha"))
+    .apply(BeamSql.fromTextRow(tableASchema));
+TableSchema tableBSchema = ...;
+PCollection<BeamSqlRow> inputTableB = p.apply(TextIO.read().from("/my/input/pathb"))
+    .apply(BeamSql.fromTextRow(tableBSchema));
+
+//run a simple query, and register the output as a table in BeamSql;
+String sql1 = "select MY_FUNC(c1), c2 from TABLE_A";
+PCollection<BeamSqlRow> outputTableA = inputTableA.apply(BeamSql.simpleQuery(sql1))
+        .withUdf("MY_FUNC", myFunc);
+
+//run a JOIN with one table from TextIO, and one table from another query
+PCollection<BeamSqlRow> outputTableB = PCollectionTuple.of(
+    new TupleTag<BeamSqlRow>("TABLE_O_A"), outputTableA)
+                .and(new TupleTag<BeamSqlRow>("TABLE_B"), inputTableB)
+    .apply(BeamSql.query("select * from TABLE_O_A JOIN TABLE_B where ..."));
+
+//output the final result with TextIO
+outputTableB.apply(BeamSql.toTextRow()).apply(TextIO.write().to("/my/output/path"));
+
+p.run().waitUntilFinish();
+ * }
+ * </pre>
+ *
+ * <h1>Beam SQL DSL vs CLI client</h1>
+ * Beam SQL CLI client(pending) is another way to leverage SQL to process data,
+ * which is built on DSL interfaces.
+ *
+ */
+public class BeamSql implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(BeamSql.class);
+
+  private static SchemaPlus schema;
+  private static BeamQueryPlanner planner;
+
+  static{
+    //Disable Assertion in Calcite temporally, it would not ignore any grammar errors.
+    ClassLoader.getSystemClassLoader().setDefaultAssertionStatus(false);
+    schema = Frameworks.createRootSchema(true);
+    planner = new BeamQueryPlanner(schema);
+  }
+
+  /**
+   * Add a UDF function.
+   */
+  public static void addUDFFunction(String functionName, Class<?> className, String methodName){
+    schema.add(functionName, ScalarFunctionImpl.create(className, methodName));
+  }
+
+  /**
+   * add a {@link BaseBeamTable} to schema repository.
+   *
+   */
+  public static void registerTable(String tableName, BaseBeamTable table) {
+    schema.add(tableName, table);
+    planner.getSourceTables().put(tableName, table);
+  }
+
+  /**
+   * Find {@link BaseBeamTable} by table name.
+   */
+  public static BaseBeamTable findTable(String tableName){
+    return planner.getSourceTables().get(tableName);
+  }
+  /**
+   * explain and display the logical execution plan.
+   */
+  public static String explainQuery(String sqlString)
+      throws ValidationException, RelConversionException, SqlParseException {
+    BeamRelNode exeTree = planner.convertToBeamRel(sqlString);
```

Comment:
That means a separated table/UDF/... namespace(aka `BeamSqlEnv`) for each query, is that expected? The existing implementation has a shared namespace, which is closer to the `session` concept.